distml/strategy/ps_strategy.py

```diff
@@ -0,0 +1,702 @@
+        num_gpus_per_server (int): number of GPUs allocated per server.
+    """
+
+    def __init__(self,
+                 *,
+                 training_operator_cls,
+                 operator_config=None,
+                 initialization_hook=None,
+                 num_workers=1,
+                 num_ps=1,
+                 num_cpus_per_worker=1,
+                 num_gpus_per_worker=1,
+                 num_cpus_per_server=1,
+                 num_gpus_per_server=1,
+                 **kwargs):
+        self.assignments = None
+
+        assert num_ps
+        self.num_ps = num_ps
+        self.num_workers = num_workers
+        self.num_cpus_per_server = num_cpus_per_server
+        self.num_gpus_per_server = num_gpus_per_server
+
+        super(ParameterServerStrategy, self).\
+            __init__(training_operator_cls=training_operator_cls,
+                     operator_config=operator_config,
+                     initialization_hook=initialization_hook,
+                     num_cpus_per_worker=num_cpus_per_worker,
+                     num_gpus_per_worker=num_gpus_per_worker,
+                     **kwargs)
+
+        # PS strategy needs some other prep-up.
+        self._init_strategy()
+
+        if operator_config and operator_config.get("batch_size"):
+            self._global_batch_size = operator_config.get("batch_size")
+        if self._global_batch_size:
+            self._collector = ThroughputCollection(
+                batch_size=self._global_batch_size)
+        else:
+            self._collector = ThroughputCollection()
+
+    def _init_strategy(self):
+        """Do initialization for the distributed strategy."""
+        # All sync with worker 0
+        init_weights_id = self.worker_group.get_named_parameters(cpu=True)
+
+        self._round_robin_sharding()
+
+        # set assignments to every worker
+        self.worker_group.set_assignments(self.assignments)
+
+        # all workers get synced
+        for i, worker in enumerate(self.worker_group.actors):
+            if i != 0:
+                ray.get([worker.set_parameters.remote(init_weights_id)])
+
+        # now spawn parameter server actors
+        shard_ids = self.worker_group.split_parameters(self.assignments)
+
+        # TODO(HUI): use scatter to send parameters
+        for server_idx, server in enumerate(self.server_group.actors):
+            this_shard_ref = self.worker_group.actors[0].index_shard.remote(
+                shard_ids, server_idx)
+            ray.get([server.set_params.remote(this_shard_ref)])
+
+    def _start_workers(self):
+        """Create worker(actor), maybe need worker group to manager these workers.
+           Or, send these workers to strategy to manager?
+
+           set workers or worker group
+           set worker info, record rank, backend, use_num_gpus?
+        """
+        # TODO (Hao): infer the per-replica batch size here...
+
+        # so here we get two set of params that will be passed around:
```

Comment:
you can clean this comment as it is redundent with those I left in `AllReduceStrategy`